tesseract/tesseract_planning/tesseract_process_planners/include/tesseract_process_planning/process_definition.h

```diff
@@ -0,0 +1,162 @@
+ */
+class ProcessTransitionGenerator
+{
+public:
+  virtual ~ProcessTransitionGenerator() = default;
+  virtual std::vector<tesseract_planners::WaypointPtr> generate(const tesseract_planners::WaypointPtr& start_waypoint,
+                                                                           const tesseract_planners::WaypointPtr& end_waypoint) const = 0;
+};
+typedef std::shared_ptr<ProcessTransitionGenerator> ProcessTransitionGeneratorPtr;
+typedef std::shared_ptr<const ProcessTransitionGenerator> ProcessTransitionGeneratorConstPtr;
+
+
+/**
+ * @class tesseract_process_planners::ProcessDefinitionConfig
+ * @details The Process Definition Config
+ *
+ * This provides the high level process configuration information. It requires the user to provide
+ * the home position waypoint (JointWaypoint) and a set of tool paths (strokes) that should be
+ * converted into a process results definition leveraging both this configuration information and
+ * the ProcessSegmentDefinitions.
+ *
+ * Also, other operations that are nice to have is the ability to offset the process. Particularly useful
+ * when wanting to verify the process without making contact with a surface in the case of sanding.
+ */
+struct ProcessDefinitionConfig
+{
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  tesseract_planners::WaypointPtr start;
+  std::vector<std::vector<tesseract_planners::WaypointPtr>> tool_paths;
+
+  std::vector<ProcessTransitionGeneratorConstPtr> transition_generator;
+
+  Eigen::Isometry3d local_offset_direction;
+  Eigen::Isometry3d world_offset_direction;
+
+  ProcessDefinitionConfig()
+  {
+    local_offset_direction.setIdentity();
+    world_offset_direction.setIdentity();
+  }
+};
+
+
+/**
+ * @brief The ProcessStepGenerator class
+ */
+class ProcessStepGenerator
+{
+public:
+  virtual ~ProcessStepGenerator() = default;
+
+  virtual std::vector<tesseract_planners::WaypointPtr> generate(const std::vector<tesseract_planners::WaypointPtr>& waypoints,
+                                                                           const ProcessDefinitionConfig& config) const = 0;
+};
+typedef std::shared_ptr<ProcessStepGenerator> ProcessStepGeneratorPtr;
+typedef std::shared_ptr<const ProcessStepGenerator> ProcessStepGeneratorConstPtr;
+
+
+/**
+ * @brief The Process Segment Definition Configuration
+ *
+ * In most manufacturing process like (sanding, painting, etc.) it includes a series of process paths (strokes).
+ * Each of the process paths may have an approach, departure and transition step.
+ *
+ * Example: In the case of painting, when approaching the part you may want to take a specific trajectory
+ * and the same for when leaving the part.
+ */
+struct ProcessSegmentDefinitionConfig
+{
+  ProcessStepGeneratorConstPtr approach;
+  ProcessStepGeneratorConstPtr process;
+  ProcessStepGeneratorConstPtr departure;
+};
+
+
+ProcessDefinition generateProcessDefinition(const ProcessDefinitionConfig& process_config,
```

Comment:
Documentation? I believe the first one applies the same segment config to all segments and  the second applies a different one to each.

Reply:
I also realize that it's not your fault that it's missing. Just making the note